### PR TITLE
Rama modulo4 new

### PR DESCRIPTION
**Pull request by oscarperezbaztan**
Para el desarrollo de esta rama he creado 4 ficheros:
- model_selection.ipynb
- train.py
- inference.py
- utils.py

**model_selection.ipynb**
Para el desarrollo de este notebook he partido como Baseline del mejor modelo del que partíamos en el módulo 3, la regresión logística con un factor de regularización 1e-4. A partir de este punto, he comenzado probando un modelo sencillo, como un árbol de decisión, para ver su comportamiento y, sobre todo, profundizar y jugar con sus híperparámetros para ver su tendencia al over-fitting, cómo controlarlo etc...
image

Una vez explorados los árboles de decisión y viendo que es un modelo realmente sencillo, he pasado a los RF. Aquí, he ido repasando los conceptos de la clase teórica antes de implementar cada modelo, para ver si tenía sentido lo que estaba haciendo. Aquí, he visto que realmente no tiene sentido aplicar técnicas de poda al árbol dado que queremos minimizar el bias (siendo el bias del algoritmo global el peor de todos), y he visto cómo se modificaba el comportamiento de este algoritmo con algunos híperparámetros. Me ha surgido la duda de si realmente entiendo correctamente la tendencia a sobre-entrenar de cada uno de los árboles. En este sentido, he restringido el número de ejemplos que se introducen en cada árbol para mejorar su independencia, y realmente el AUC en train es coherente con el de val, pero no sé si realmente esta práctica tiene sentido hacerla.
image

Una vez vistos los RF, he pasado a los Gradient Boosting, y los resultados han mejorado con respecto a los modelos usados previamente. De nuevo, he intentado jugar con los híperparámetros de forma ordenada para ver cómo afectaba el comportamiento del algoritmo en general:
image

Por último, he probado con el XGBoost, sí que es verdad que en este caso me ha costado más obtener un punto de partido en cuánto a híperparámetros dada la amplia variedad de grados de libertad que contiene este modelo. Finalmente, he ido probando diferentes casos y me he quedado con este modelo de forma definitiva aunque sí que es verdad que sus prestaciones con respecto a Gradient Boosting o RF son muy similares:
image

**utils.py**

Por lo que respecta a este fichero, he metido algunas funciones para evitar repetir información en el notebook y que no sea redundante, así como algunas formas de optimizar el código para descargar los datos, realizando de forma automática el muestreo etc.

**train.py and inference.py**

En cuánto a estos dos ficheros, he probado a implementarlo de la forma en la que se hizo en el módulo 3. Sí que es verdad que me hubiera gustado dedicarle mucho más tiempo a esta parte porque creo que es más importante que el notebook y me gustaría profundizar más en ella, aplicando el diccionario que se proponía e intentando hacerla mejor pero sí que quería adjuntarla en esta PR para ver si he hecho correctamente esa primera idea inicial.

**Resumen**

En general este módulo me ha permitido hacer mucho research en cuanto a la parte de modelos no lineales, intentando entenderlos mejor, cómo se comportan y, sobre todo, intentando aplicar ese conocimiento teórico a la práctica. De la misma manera, he intentado aplicar esos comentarios del módulo 3 y mejorando ciertos aspectos, como no usar cv para la selección del modelo por el data leakage etc. En cualquier caso, como he comentado, sí que me gustaría profundizar bastante más en toda la parte de implementación del modelo y la realización del pipeline, por lo que mi intención es, en base a los comentarios de esta primera pr, intentar profundizar sobre todo en esta parte.

### CODE REVIEW

**Comment by guillerbf** [comment]
Comentarios generales antes de ver el código:
1. Las PR así dan gusto. Muy buena descripción de lo que has hecho y porqué. Ayuda muchiiisimo al revisor a entender la lógica detrás de lo que has hecho.
2. En cuanto a hiperparámetros de RF. La realidad es que en un random forest, el error de entrenamiento debe ser 0 ([mira esto](https://zrive-ds.slack.com/archives/C06EXLAD532/p1708613412257899)) por diseño y si los datos son realmente independientes y las variables tbn, en teoría no habría que tocar ningun hiperparametro. La realidad no siempre es así y por eso puede tener sentido hp_tuning.
3. Xgboost y GB no son cosas distintas. Son implementaciones distintas del mismo concepto. No familias distintas de algoritmos.

File: src/module_4/train.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+TRAIN_SIZE = 0.7
+
+OUTPUT_PATH = os.path.join(STORAGE_PATH, "module_4_models")
+
+PARAM_GRID = {'learning_rate': [0.001, 0.01, 0.1, 0.25, 1, 10]}
+
+
+def train_test_split(df:pd.DataFrame, train_size:float) -> Tuple[pd.DataFrame, pd.Series, pd.DataFrame, pd.Series]:
+
```
Comment (guillerbf) [inline src/module_4/train.py:39]: This will benefit from a different name and or some docstrings. `train_test_split` is a very well-known sklearn method that could be confusing since you are splitting data here based on time and using quantiles.

File: src/module_4/train.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        precision, recall, _ = precision_recall_curve(y_val, y_pred)
+        pr_auc = auc(recall, precision)
+
+        if pr_auc > best_auc:
+            best_auc = pr_auc
+            best_params = {'learning_rate': learning_rate}
+
+    logger.info(f"Training best model with learning_rate = {best_params['learning_rate']} over whole dataset")
+    best_model = xgb.XGBClassifier(
```
Comment (guillerbf) [inline src/module_4/train.py:101]: DRY principle. Most of this call is repeated? Are there parameters that are the default ones and therefore could be ignored?

Otherwise, can you add a `class XGBoostModel` with a constructor with most of those default values and a docstring explaining that they come from the research you did. Good practice to add a link to notebook in those cases, so the reviewer or anyone in the future can go and look how those parameters where found.

File: src/module_4/train.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+        xgboost.fit(X_train, y_train)
+
+        y_pred = xgboost.predict_proba(X_val)[:, 1]
+
+        precision, recall, _ = precision_recall_curve(y_val, y_pred)
+        pr_auc = auc(recall, precision)
+
+        if pr_auc > best_auc:
+            best_auc = pr_auc
+            best_params = {'learning_rate': learning_rate}
```
Comment (guillerbf) [inline src/module_4/train.py:98]: This is not quite right. Is the model with largest `pr_auc` the best model in every case? 

I'd add a comment at least saying "since we don't have any more information from the business to understand what precision-recall requirements they have, we are assuming that greater pr_auc is better"

File: src/module_4/inference.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+from utils import *
+
+logger = logging.getLogger(__name__)
+logger.level = logging.INFO
+
+consoleHandler = logging.StreamHandler()
+logger.addHandler(consoleHandler)
+
+def main():
+    model_name = "20240217-201103_XGBoost_0.25.pkl"
```
Comment (guillerbf) [inline src/module_4/inference.py:14]: This should be a config. You can set it up as a global variable but ideally this script should be called passing which model to load, right?

File: src/module_4/inference.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
+import logging
+import os
+from joblib import load
+from train import *
```
Comment (guillerbf) [inline src/module_4/inference.py:4]: explicit imports >>>> than wildcards

File: src/module_4/inference.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+consoleHandler = logging.StreamHandler()
+logger.addHandler(consoleHandler)
+
+def main():
+    model_name = "20240217-201103_XGBoost_0.25.pkl"
+    model = load(os.path.join(OUTPUT_PATH, model_name))
+    logger.info(f"Loaded model {model_name}")
+
+    df = preprocess_data_manually()
```
Comment (guillerbf) [inline src/module_4/inference.py:18]: With wildcard import I don't know where this function comes from

File: src/module_4/inference.py [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    X, y = feature_label_split(FEATURE_COLS, TARGET,df)
+
+    y_pred = model.predict_proba(X)[:,1]
+
+    evaluate_model("Inference test", y, y_pred)
```
Comment (guillerbf) [inline src/module_4/inference.py:23]: An inference pipeline isn't to evaluate a model but to serve predictions for new data.

Typically something like:

```
MODEL = load_model(model_path)

def predict(x):
    if not validate_input(x): # does it has all the columns needed and so.
        raise error
    return MODEL.predict_proba(x)[1]
```

If you see Pablo's video you will see that he uses classes since they are a very good fit. You can create a Model class which in fact inherits from sklearn base learner and you overwrite the methods `.fit`, `.predict_proba` and `predict` if you want.

In `train.py` you instantiate such a class` do the `.fit` and save it.

In `inference.py` load the trained model and use `.predict_proba` or `predict` depending on the use case to make predictions for new data.

File: src/module_4/utils.py [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+
+
+def relevant_orders(df:pd.DataFrame, min_products: int = 5) -> pd.DataFrame:
+    num_products_ordered = df.groupby('order_id')['outcome'].sum()
+    orders_with_5_or_more_unique_products = num_products_ordered[num_products_ordered>=min_products].index
+    filtered_data = df[df['order_id'].isin(orders_with_5_or_more_unique_products)]
+    return filtered_data
+
+
+def preprocess_data_manually() -> pd.DataFrame:
```
Comment (guillerbf) [inline src/module_4/utils.py:46]: Why this name? What's the manual process here?

**Review by guillerbf — COMMENTED** [review]
Good PR.

Take a look at how to structure code in `train` and `inference`. Apart from that every other comment is more about cleanliness and maintainability that functional issues.
